hsakmt: Expose and use CWSR and Control stack sizes

kentrussell opened the pull request:
This will allow VGPR size to be determined dynamically, instead of relying on hardcoding

## Motivation

<!-- Explain the purpose of this PR and the goals it aims to achieve. -->

## Technical Details

<!-- Explain the changes along with any relevant GitHub links. -->

## JIRA ID

<!-- If applicable, mention the JIRA ID resolved by this PR (Example: Resolves SWDEV-12345). -->
<!-- Do not post any JIRA links here. -->

## Test Plan

<!-- Explain any relevant testing done to verify this PR. -->

## Test Result

<!-- Briefly summarize test outcomes. -->

## Submission Checklist

- [ ] Look over the contributing guidelines at https://github.com/ROCm/ROCm/blob/develop/CONTRIBUTING.md#pull-requests.

kentrussell committed:
hsakmt: Expose CWSR and Control stack sizes

This is better than hardcoding values and hoping that they align with
KFD's definitions

Signed-off-by: Kent Russell <kent.russell@amd.com>

superm1 commented:
Mostly to connect dots for anyone who looks at this later: this matches [this kernel change](https://lore.kernel.org/amd-gfx/20251205184158.2261400-1-mario.limonciello@amd.com/).

dayatsin-amd reviewed (changes requested):
In thunk, when CwsrSize and CtlStackSize are not zero, we need to add code to use these values instead of trying to calculate them.

This would be in queues.c:update_ctx_save_restore_size

kentrussell commented:
Right. That felt like it should be a separate change, and I wasn't confident enough in knowing where all those calculations had to happen

jokim-amd commented:
"This will allow VGPR size to be determined dynamically, instead of relying on hardcoding"
If the purpose of this change is just to expose the KFD authority save area size internally to Thunk with the idea that a follow on patch will actually use mentioned authority, then the description should mention that (also just mention the overall size, no one really needs to know about VGPRs).

But I don't see a need to do this unless we're going to use actually use it during queue creation.

jokim-amd commented:
"In thunk, when CwsrSize and CtlStackSize are not zero, we need to add code to use these values instead of trying to calculate them."

I think we should still calculate this in Thunk for existing products.
The KFD is also broken for GFX1151 in reporting the correct size (patch in flight to correct this so have no version control because it's a bug).
Also user space should be allowed to over allocate if it wants.
I think for any existing chips which have been relying on thunk should max(kfd_cwsr_size, thunk_cwsr_size).
Future chips currently in npi can just rely on kfd_cwsr_size.

kentrussell commented:
> "This will allow VGPR size to be determined dynamically, instead of relying on hardcoding" If the purpose of this change is just to expose the KFD authority save area size internally to Thunk with the idea that a follow on patch will actually use mentioned authority, then the description should mention that (also just mention the overall size, no one really needs to know about VGPRs).
> 
> But I don't see a need to do this unless we're going to use actually use it during queue creation.

Fair enough. I didn't know who would know which part needed to be done next, but I can reword the commit message accordingly. This isn't my area of expertise. But I can try to throw something together

kentrussell committed:
hsakmt: Use CwsrSize and CtlStackSize if available

If KFD is providing the CwsrSize and CtlStackSize, use the maximum
of those and the old calculations for the ctx_save_restore_size
and ctl_stack_size defined in the queue

Signed-off-by: Kent Russell <kent.russell@amd.com>

jokim-amd commented on projects/rocr-runtime/libhsakmt/src/queues.c:
```diff
@@ -343,8 +343,11 @@ static bool update_ctx_save_restore_size(HsaKFDContext *ctx, uint32_t nodeid, st
+		/* Keep calculating it in case we are using an older kernel, but if we have
+		 * the CtlStackSize from KFD, use the max of that and the calculation
+		 */
+		q->ctx_save_restore_size = MAX(node.CtlStackSize, q->ctl_stack_size
```
i think you need to do  actually cwsr area size (stack + wg_context) if the kfd has it

superm1 replied:
Something we talked about was emitting something to stderr to tell people to update kernels too.  Is this the right place to do it?

kentrussell replied:
We can do that if we want to be explicit. This is just "hey, we're using the stuff from KFD if it exists" . If we want to go the route of "you really should update your kernel" we can throw that too . The biggest concern would be that for every queue that gets made, it'll throw it. Which will get pretty noisy pretty quickly. I guess that's where the ABI check can come in

jokim-amd replied:
the existence of control stack & cwsr area size properties guarantees proper "ABI" restoration i.e. kernel allows non-fatal user space override of save area (reminder - we should probably also check for KFD control stack size hints as well and use them if they exist).

the only case we would care about printing is on queue creation failure for gfx1151 when those properties don't exist, in which the explanation (print) for the failure "may" be due to the kernel's old bad restriction check.

we can't really print ahead of time because we'd be making a lot of overkill assumptions on what a "working" sw stack is during runtime i.e. we really only needed the one kernel patch that relaxed cwsr area size over allocation & acquire the get this version of runtime working and the only way to guess is if the kernel has the minimal fix in the absence of mentioned properties, is if queue creation is fine with this version of thunk.

i'd say print if -> gfx1151 && !kfd_cwsr_area_size_hint && queue creation fails (maybe debug print this with the existing thunk debug envar if spam verbose-ness is an issue)
that way, user space will be nice and silent during normal ops and with minimal changes required to make this work.

superm1 replied:
The only problem with that approach is that these failures lead to hangs/reboots.  So you want to be upfront with people that they should upgrade their kernel to pick up the fixes for a better experience.

jokim-amd replied:
queue failure wouldn't lead to system hangs or reboots because the queue would fail to be created in the first place if user space is error handling correctly, so we should just get a fatal error at the app level.

if we want to surface this ahead of time, then that sounds like something ROCr (@dayatsin-amd ) would have to consider during HSA init.  and again we would be printing out warnings that users "may" need to upgrade their kernel (if we're okay with that) in the absence of the kfd sizing hints.

kentrussell replied:
I'm working on a patch to print it during topology init, if it's GFX1151+ABI<20. Just saying "hey, you're on GFX1151 and KFD's ABI is <1.20. You should update KFD to avoid issues".

Just need to try to test it out. But we could let ROCr do it too, if desired.

jokim-amd replied:
yeah i guess we could put it in thunk when we acquire system properties.  looks like we only call that once during rocr init anyways.

dayatsin-amd replied:
This should be in Thunk acquire system properties. The place where we print should be "close" to where we do the checks.

kentrussell committed:
hsakmt: Add warning when ABI<1.20 on GFX1151

CwsrSize and CtlStackSize are reported by KFD ABI 1.20. GFX1151
specifically may have some issues if these regions are misaligned, so
report a strong warning during topology initialization if the system is
GFX1151 but is using KFD ABI < 1.20

Signed-off-by: Kent Russell <kent.russell@amd.com>
